rtgui/compressgamut.cc

```diff
@@ -0,0 +1,366 @@
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with RawTherapee.  If not, see <https://www.gnu.org/licenses/>.
+ */
+ /*
+ * tweaked from the original from https://github.com/jedypod/gamut-compress
+ * https://docs.acescentral.com/specifications/rgc/
+*/ 
+#include "compressgamut.h"
+
+#include "eventmapper.h"
+
+#include "../rtengine/procparams.h"
+
+using namespace rtengine;
+using namespace rtengine::procparams;
+
+const Glib::ustring Compressgamut::TOOL_NAME = "compressgamut";
+
+Compressgamut::Compressgamut () : FoldableToolPanel(this, TOOL_NAME, M("TP_COMPRESSGAMUT_LABEL"), false, true)
+{
+    auto m = ProcEventMapper::getInstance();
+    EvcgColorspace = m->newEvent(COMPR, "HISTORY_MSG_CG_COLORSPACE");//FIRST to reinit histogram
+    Evcgthc = m->newEvent(COMPR, "HISTORY_MSG_CG_CYANTH");
+    Evcgthm = m->newEvent(COMPR, "HISTORY_MSG_CG_MAGENTATH");
+    Evcgthy = m->newEvent(COMPR, "HISTORY_MSG_CG_YELLOWTH");
+    Evcgdc = m->newEvent(COMPR, "HISTORY_MSG_CG_CYANLIM");
+    Evcgdc = m->newEvent(COMPR, "HISTORY_MSG_CG_CYANLIM");
+    Evcgdc = m->newEvent(COMPR, "HISTORY_MSG_CG_CYANLIM");
+    Evcgdm = m->newEvent(COMPR, "HISTORY_MSG_CG_MAGENTALIM");
+    Evcgdy = m->newEvent(COMPR, "HISTORY_MSG_CG_YELLOWLIM");
+    Evcgroll = m->newEvent(COMPR, "HISTORY_MSG_CG_ROLLOFF");
+    Evcgpwr = m->newEvent(COMPR, "HISTORY_MSG_CG_VALUE");
+    Evcgenabled = m->newEvent(COMPR, "HISTORY_MSG_CG_ENABLED");
+
+
+
+    Gtk::Frame *iFrame = Gtk::manage(new Gtk::Frame(M("TP_COMPRESSGAMUT_MAIN_COLORSPACE")));
+
+    iFrame->set_label_align(0.025, 0.5);
+    iFrame->set_tooltip_markup (M("TP_COMPRESSGAMUT_COLORSPACE_TOOLTIP"));
+
+    Gtk::Box *iVBox = Gtk::manage(new Gtk::Box(Gtk::ORIENTATION_VERTICAL));
+   
+    colorspace = Gtk::manage(new MyComboBoxText());
+    colorspace->append(M("TP_COMPRESSGAMUT_REC2020"));
+    colorspace->append(M("TP_COMPRESSGAMUT_PROPHOTO"));
+    colorspace->append(M("TP_COMPRESSGAMUT_ADOBE"));
+    colorspace->append(M("TP_COMPRESSGAMUT_SRGB"));
+    colorspace->append(M("TP_COMPRESSGAMUT_DCIP3"));
+    colorspace->append(M("TP_COMPRESSGAMUT_ACESP1"));
+    colorspace->set_active(3);
+    iVBox->pack_start(*colorspace);
+    iFrame->add(*iVBox);
+    pack_start(*iFrame);
+    colorspaceconn = colorspace->signal_changed().connect(sigc::mem_fun(*this, &Compressgamut::colorspaceChanged));
+
+    // Percentage of the core gamut to protect Limits
+    // Values calculated to protect all the colors of the ColorChecker Classic 24 as given by
+    // ISO 17321-1 and Ohta (1997)
+
+    th_c = Gtk::manage (new Adjuster (M("TP_COMPRESSGAMUT_CYANTH"), 0., 0.999, 0.001, 0.815));//0.999 to avoid: 1 - th = 0
+    th_m = Gtk::manage (new Adjuster (M("TP_COMPRESSGAMUT_MAGENTATH"), 0., 0.999, 0.001, 0.803));
+    th_y = Gtk::manage (new Adjuster (M("TP_COMPRESSGAMUT_YELLOWTH"), 0., 0.999, 0.001, 0.880));
```

Comment:
Is there really a problem with a threshold of 1? The equations look like they are fine with it. The only problem I see is when threshold is equal to the limit, but the limit is always greater than 1.

Reply:
I noticed - especially on high gamut reds - that 1 brought artifacts
Here 
    for (int i = 0; i < 3; i = i+1) {
        s[i] = (1.0f  - threshold[i]) / sqrt(fmax(1.001f, distance_limit[i]) - 1.0f);

1.0f - - threshold[i]  becomes null

Reply:
No problem on principle... In some images (red flowers) artefacts.